lib/charms/microcluster_token_distributor/v0/token_distributor.py

```diff
@@ -115,6 +174,108 @@ def _call_cluster_command(self, *args):
         )
         return result.returncode, result.stdout
 
+    def _to_mirror_key(self, key):
+        return "{0}{1}".format(MIRROR_PREFIX, key)
+
+    def add_to_mirror(self, relation, data: dict[str, str]):
+        """Add data into the units databag in a mirrorable form."""
+        for k, v in data.items():
+            relation.data[self.charm.unit][self._to_mirror_key(k)] = v
+
+    def find_value(self, relation, key, keep_empty=True):
+        """Find corresponding value in mirror.
+
+        Return false if value not found, if keep_empty is enabled 'empty' an
+        empty value will be accepted.
+        """
+        values = [
+            value
+            for mirror in self.find_mirrors(relation)
+            if (value := relation.data[mirror].get(self._to_mirror_key(key)))
+            and (keep_empty or value != "empty")
+        ]
+
+        if len(values) == 0:
+            return False
+
+        return corroborate(values, default="empty")
+
+    def any_data_exists(self, relation):
+        """Check if there is any non empty value on the remote side of the mirror."""
+        for unit in relation.units:
+            for k in relation.data[unit].keys():
+                if k.startswith(MIRROR_PREFIX) and relation.data[unit][k] != "empty":
+                    return True
+        return False
+
+    def find_mirrors(self, relation) -> list[ops.Unit]:
+        """Find all remote mirror units."""
+        relation_data = relation.data
+        distributor_mirrors = [
+            unit
+            for unit in relation.units
+            if (mirror := relation_data[unit].get("mirror")) and mirror == "up"
+        ]
+        return distributor_mirrors
+
+    def get_relevant_mirror_data(self, relation, keep_empty=True) -> dict[str, str]:
```

Comment:
nit: I love type hints for return values. It'd be nice to have them consistently everywhere